kullanıcının bize verdiği sayıyla, o sayının karakökünü veya karesini bulan bir program yapıyoruz

In [2]:
veri = input("Karakökünü hesaplamak istediğiniz sayı: ")
karakok = int(veri) ** 0.5
print(veri,"sayısının karakökü: ",karakok)

Karakökünü hesaplamak istediğiniz sayı: s


ValueError: ignored

kullanıcı eğer sayı değerli bir veri girmek yerine string bir ifade girerse hata alırız.

işte bunun gibi hatalar almamak için, hataları önceden kestirip çözmemiz lazım.

# Hata Türleri

+ programcı hataları (Error)
+ program kusurları (Bug)
+ İstisnalar (Exception)

**Programcı Hataları**

Programı yazan kişinin dikkatsizliğinden dolayı ortaya çıkan hatalardır:

In [4]:
print "Merhaba Dünya"

SyntaxError: ignored

SyntaxError, söz dizimi hatası demektir
bu hatalar proglama diline ilişkin bir özelliğin kullanımından yani programcının yaptığı yazım yanlışlarından kaynaklanır

Bu tür hataların ise tespit edilmesi daha kolaydır

**Problem Kusurları (Bug)**

diğer hataların içinde tespit edilmesi en zor olanıdır. Çünkü bu kusurlu programlar hata vermeden çalışır. 

örneğin bir programda formül hatası yaptık, program belirlediğimiz formüle uyar ve sonucu gösterir, fakat formül hatası yaptığımız için sonuçlar yanlış çıkar:

In [5]:
number1 = input("Toplama işlemi için ilk sayıyı giriniz: ")
number2 = input("Toplama işlemi için ikinci sayıyı giriniz: ")
print(number1,"+",number2,"=",number1+number2)

Toplama işlemi için ilk sayıyı giriniz: 24
Toplama işlemi için ikinci sayıyı giriniz: 4
24 + 4 = 244


24 ile 4 sayısının toplamını 28 bulacağımız yerinde programımızda bug olduğundan dolayı program toplamak yerine birleştirip bize 244 sonucunu gösterdi.

Bu yüzden düzgün çalıştığını düşündüğümüz programlarda bug olmuş olabilir, o yüzden dikkat etmek lazım.

mesala eval() fonskiyonunu dikkatsizce  kullanıldığını zaman güvenlik kusuru yaratır

buna ise security bug veya security flaw denir.

**İstisnalar (Exception)**

Programın çalışması sırasında ortaya çıkan istisnai, çok fazla görülmeyen hatalardır.

In [6]:
ilkSayi = input("İlk sayı: ")
ikinciSayi = input("ikinci sayı: ")
ilkSayi = int(ilkSayi)
ikinciSayi = int(ikinciSayi)
print(ilkSayi,"/",ikinciSayi,"=",ilkSayi/ikinciSayi)

İlk sayı: ss
ikinci sayı: 20


ValueError: ignored

kullanıcı burada sayı yerine string ifadeli veri girmiştir
burada ki hata ise int() fonskiyonun string ifadeyi int veri tipine dönüştürememisidir.

Veya kullanıcı bir sayıyı 0 a bölmeye çalışadabilir

In [8]:
numberOne = input("İlk sayı: ")
numberTwo = input("İkinci sayı: ")
numberOne = int(numberOne)
numberTwo = int(numberTwo)
print(numberOne,"/",numberTwo,"=",numberOne/numberTwo)

İlk sayı: 25
İkinci sayı: 0


ZeroDivisionError: ignored

kullanıcı bir sayıyı 0 a bölmeye çalışında hata aldık, çünkü matematikte sayılar 0'a bölünemez

ValueError ve ZeroDivisionError hataları istisna hatalara örnektir.